### DISCUSSION TITLE
Please help to write JavaScript function to convert old configuration to 12 - Merlin

### DISCUSSION THREAD

**Discussion by mario-website** [RavenSystem/esp-homekit-devices#2031] [Q&A]
Is this even possibile to do it?
Could anyone help me to write a function in JavaScript who will convert old configuration to new one?

**Comment by mario-website** [comment]
Ok, I started and did this:
https://converter-to-merlin.netlify.app
basically is working only for case if using from https://glumb.github.io/haa-configurator/ SONOFF 4CH config.

**Comment by mario-website** [comment]
Converter is growing but I need YOUR help. 
This converter works for now ONLY IF some condition is meet. An working example is to convert from Sonoff 4CH template form https://glumb.github.io/haa-configurator/.
Few minutes ago I just added a condition to include "l" value in array where is [[...allOtherValues],2]
...there is so many condition to check in conversion but probably, one day all of them will be checked and converter will be working as expected.
**HOW YOU CAN HELP?**
Please, if this website has helped anyone, share your code. I can use it in the testing area as an example of input and expected output. In the future, when this converter passes new and more conditions, previously working conditions might unintentionally fail.
example:
input:
`{"c":{"l":12,"n":"HAA-CAA776","h":1,"b":[{"g":3,"t":5}]},"a":[{"t":2,"0":{"r":[{"g":14,"v":0},{"g":13,"v":1}]},"1":{"r":[{"g":14,"v":1},{"g":13,"v":0}]},"b":[{"g":3,"t":1}]}]}`
output:
`{"c":{"l":12,"n":"HAA-CAA776","h":1,"b":[{"g":3,"t":5}],"io":[[[12,14,13],2],[[3],6]]},"a":[{"0":{"r":[{"g":14,"v":0},{"g":13,"v":1}]},"1":{"r":[{"g":14,"v":1},{"g":13,"v":0}]},"t":2,"b":[{"g":3,"t":1}]}]}`

**Comment by mario-website** [comment]
I just released a new version, could anyone to check how is working now?
@RavenSystem could you take a look and give me a feedback?

**Reply by JonnyFPV** [reply]
Hi, thank you for your work.

I tried your converter because I don't understand the new cheme.

Some parts Working now.

but 1 problem I got for now. It's a simple relay Board from eBay.

its for my door lock. So with your converter the Board is triggering the relay ok. the boot status I correct. But I declared a timer to set it back to (closed) position. With your converter the relay get triggered by HomeKit and reported status ok (open). then if the timer is up the status In HomeKit got back to closed status but the relay does not turn off. 

Old: 
`{
  "c": {
    "l": 13,
    "b": [
      {
        "g": 0,
        "t": 5
      }
    ]
  },
  "a": [
    {
      "0": {
        "r": [
          {
            "g": 12,
            "v": 1,
            "i": 5
          }
        ]
      },
      "1": {
        "r": [
          {
            "g": 13,
            "v": 0
          }
        ]
      },
      "t": 4,
      "b": [],
      "s": 1,
      "i": 5
    }
  ]
}`

New:

`{
  "c": {
    "l": 13,
    "b": [
      [
        0,
        5
      ]
    ],
    "io": [
      [
        [
          13
        ],
        3
      ],
      [
        [
          0
        ],
        6
      ],
      [
        [
          12,
          13
        ],
        2
      ]
    ]
  },
  "a": [
    {
      "0": {
        "r": [
          [
            12,
            1
          ]
        ]
      },
      "1": {
        "r": [
          [
            13,
            0
          ]
        ]
      },
      "t": 4,
      "b": [],
      "s": 1,
      "i": 5
    }
  ]
}`

Can anyone help to make the relay to stop after 5 seconds?

**Reply by mario-website** [reply]
Hi, I having also troubles to understand but maybe one day…
Anyway, my converter, at this moment, not convert your needs. But after reading this:

https://github.com/RavenSystem/esp-homekit-devices/wiki/Migration-Guide-for-HAA-V12-Merlin#example-3-1

remove from [ [ 12, 13 ], 2 ] ] number 13 and add new array with  [ [ 13 ], 6,5 ] ] or  [ [ 13 ], 6,0,5 ] ]
I also removed [ [ 13 ], 3 ] but still don’t understand what to do in case having “l”:13 in “c” and having “g”:13 anywhere in “c” so try with [ [ 13 ], 3 ] or without.
I suggest to to try:
{ "c": { "l": 13, "b": [ [ 0, 5 ] ], "io": [ [ [ 0 ], 6 ], [ [ 12 ], 2 ] , [ [ 13 ], 6,5 ] ] }, "a": [ { "0": { "r": [ [ 12, 1 ] ] }, "1": { "r": [ [ 13, 0 ] ] }, "t": 4, "b": [], "s": 1, "i": 5 } ] }
Or
{ "c": { "l": 13, "b": [ [ 0, 5 ] ], "io": [ [ [ 0 ], 6 ], [ [ 12 ], 2 ] , [ [ 13 ], 6,0,5 ] ] }, "a": [ { "0": { "r": [ [ 12, 1 ] ] }, "1": { "r": [ [ 13, 0 ] ] }, "t": 4, "b": [], "s": 1, "i": 5 } ] }

**Reply by JonnyFPV** [reply]
thank you for your reply.

the first one does the same as the provided one of your converter. Bootup status HK closed, relay off,  HomeKit button pressed: Relay get triggered, HK status open, relay is on. 5 seconds later, HK status closed, relay still remains on. 

the second one: relay and status always open/on if I press button in HomeKit..

so does not work as aspect.  :-(

**Reply by JonnyFPV** [reply]
@mario-website this code now works as aspected.

`{ "c":{ "io":[[[12],2],[[5],6,0,1]] }, "a":[{ "t":4, "i":4, "0":{"r":[[12,1,4]]}, "1":{"r":[[12]]}, "b":[[5,0]] }] }`

**Comment by RavenSystem** [comment]
New HAA Configurator: https://haaconfig.github.io/meplhaa-configurator/
